ai_flow_plugins/blob_manager_plugins/oss_blob_manager.py

```diff
@@ -81,8 +87,53 @@ def download_project(self, workflow_snapshot_id, remote_path: Text, local_path:
             repo_path = Path(tempfile.gettempdir())
         local_zip_file_path = str(repo_path / local_zip_file_name) + '.zip'
         extract_path = str(repo_path / local_zip_file_name)
-        self.bucket.get_object_to_file(oss_object_key, filename=local_zip_file_path)
+
+        if not os.path.exists(local_zip_file_path):
+            logger.debug("{} not exist".format(local_zip_file_path))
+            lock_file_path = os.path.join(repo_path, "{}.lock".format(local_zip_file_name))
+            lock_file = open(lock_file_path, 'w')
+            logger.debug("Locking file {}".format(lock_file_path))
+            fcntl.flock(lock_file.fileno(), fcntl.LOCK_EX)
+            logger.debug("Locked file {}".format(lock_file_path))
+            try:
+                if not os.path.exists(local_zip_file_path):
+                    logger.info("Downloading oss object: {}".format(oss_object_key))
+                    self._get_oss_object(local_zip_file_path, oss_object_key)
+            except Exception as e:
+                logger.error("Failed to download oss file: {}".format(oss_object_key), exc_info=e)
+            finally:
+                fcntl.flock(lock_file.fileno(), fcntl.LOCK_UN)
```

Comment:
I'm afraid the lock would be held too long.

Reply:
What do you mean? I think it is necessary to hold the lock until the file is downloaded successfullyly.